app/cloud.go

```diff
@@ -16,34 +17,93 @@ import (
 	"github.com/mattermost/mattermost-server/v6/shared/i18n"
 	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+	"github.com/mattermost/mattermost-server/v6/store"
 )
 
-func (a *App) NotifySystemAdminsToUpgrade(c *request.Context, currentUserTeamID string) *model.AppError {
-	userId := c.Session().Id
+const lastTrialNotificationTimeStamp = "LAST_TRIAL_NOTIFICATION_TIMESTAMP"
+const lastUpgradeNotificationTimeStamp = "LAST_UPGRADE_NOTIFICATION_TIMESTAMP"
+const defaultCloudNotifyAdminCoolOffDays = 0.0001157407407 // this is a temp change
 
-	fakeId := strings.ReplaceAll(model.CloudNotifyAdminInfo, "_", "") + "123456"
+// Ensure cloud service wrapper implements `product.CloudService`
+var _ product.CloudService = (*cloudWrapper)(nil)
 
-	// check if already notified
-	notificationPref, err := a.Srv().Store.Preference().Get(fakeId, model.PreferenceCloudUserEphemeralInfo, model.CloudNotifyAdminInfo)
+// cloudWrapper provides an implementation of `product.CloudService` for use by products.
+type cloudWrapper struct {
+	cloud einterfaces.CloudInterface
+}
+
+func (a *App) SaveAdminNotification(userId string, notifyData *model.NotifyAdminToUpgradeRequest) *model.AppError {
+	requiredFeature := notifyData.RequiredFeature
+	requiredPlan := notifyData.RequiredPlan
+	trial := notifyData.TrialNotification
+
+	if a.UserAlreadyNotifiedOnRequiredFeature(userId, requiredFeature) {
+		return model.NewAppError("app.SaveAdminNotification", "api.cloud.notify_admin_to_upgrade_error.already_notified", nil, "", http.StatusForbidden)
+	}
+
+	_, appErr := a.SaveAdminNotifyData(&model.NotifyAdminData{
+		UserId:          userId,
+		RequiredPlan:    requiredPlan,
+		RequiredFeature: requiredFeature,
+		Trial:           trial,
+	})
+	if appErr != nil {
+		return appErr
+	}
+
+	return nil
+}
+
+func (a *App) DoCheckForAdminNotifications(trial bool) *model.AppError {
+	ctx := request.EmptyContext(a.Srv().Log())
+	subscription, err := a.Cloud().GetSubscription("")
 	if err != nil {
-		mlog.Warn("Unable to get preference cloud_user_ephemeral_info", mlog.Err(err))
+		return model.NewAppError("DoCheckForAdminNotifications", "app.notify_admin.send_notification_post.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	if notificationPref != nil {
-		info := &model.AdminNotificationUserInfo{}
-		err = json.Unmarshal([]byte(notificationPref.Value), info)
-		if err != nil {
-			mlog.Warn("Unable to Unmarshal", mlog.Err(err))
+	products, err := a.Cloud().GetCloudProducts("", true)
+	if err != nil {
+		return model.NewAppError("DoCheckForAdminNotifications", "app.notify_admin.send_notification_post.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	currentProduct := getCurrentProduct(subscription.ProductID, products)
+	if currentProduct == nil {
+		return model.NewAppError("DoCheckForAdminNotifications", "app.notify_admin.send_notification_post.app_error", nil, "current product is nil", http.StatusInternalServerError)
```

Comment:
This is good for now, but thinking on what is coming to expand this to self-managed, wondering if we should check here for license sku instead of subscription sku. This will disassociate this feature from cloud and we can move this out of this `/cloud.go` 

(Not a blocker for this PR but can be done here if we don't want to change it later)

Reply:
Agreed